### PR TITLE
Add memoization to `Cache`

### PR DESCRIPTION
**Pull request by Jym77**

### CODE REVIEW

File: packages/alfa-aria/src/dom/predicate/is-programmatically-hidden.ts [hunk 1]
```diff
@@ -34,39 +34,32 @@ export function isProgrammaticallyHidden(
...
+    (node: Node) =>
+      node.parent(Node.fullTree).some(_hasHiddenAncestors(device, context)),
+  );
 }
+
+const hasHiddenAncestors = Cache.memoize<
+  unknown,
+  [Device, Context],
+  Predicate<Node>
+>(_hasHiddenAncestors);
```
Comment (Jym77) [inline packages/alfa-aria/src/dom/predicate/is-programmatically-hidden.ts:65]: I'm not sure why TS needs some skaffolding here, but not in the simple cases in the tests 🤷

File: packages/alfa-aria/src/dom/predicate/is-programmatically-hidden.ts [hunk 2]
```diff
@@ -34,39 +34,32 @@ export function isProgrammaticallyHidden(
   );
 }
 
-const cache = Cache.empty<Device, Cache<Context, Cache<Node, boolean>>>();
-
-function hasHiddenAncestors(
+function _hasHiddenAncestors(
```
Comment (Jym77) [inline packages/alfa-aria/src/dom/predicate/is-programmatically-hidden.ts:37]: This is just a demo of using the new decorator, not directly related to the changes.

File: packages/alfa-aria/src/dom/predicate/is-programmatically-hidden.ts [hunk 3]
```diff
@@ -34,39 +34,28 @@ export function isProgrammaticallyHidden(
...
+          device,
+          context,
         ),
-      );
+        hasAttribute("aria-hidden", equals("true")),
+      ),
+    ),
+    // Or its parent is programmatically hidden
+    (node: Node) =>
+      node.parent(Node.fullTree).some(_hasHiddenAncestors(device, context)),
```
Comment (rcj-siteimprove) [inline packages/alfa-aria/src/dom/predicate/is-programmatically-hidden.ts:57]: **Question:** Is it on purpose the call is to `_hasHiddenAncestors` and not `hasHiddenAncestors`? I can't tell if it makes a difference, but in the original code, in the recursion the cache would be used, but now it's going to recurse without the cache, I think.

Comment (Jym77) [reply packages/alfa-aria/src/dom/predicate/is-programmatically-hidden.ts:57]: Ouch. Good point. 🤔

**Review by rcj-siteimprove — APPROVED** [review]
Nice 👏🏼

**Pull request merged**